flocker/dockerplugin/test/test_api.py

```diff
@@ -8,16 +8,110 @@
 
 from twisted.web.http import OK
 from twisted.internet import reactor
+from twisted.internet.task import Clock
+from twisted.python.components import proxyForInterface
+from twisted.internet.interfaces import IReactorTime
+
+from pyrsistent import pmap
+
+from characteristic import attributes
 
 from .._api import VolumePlugin, DEFAULT_SIZE
 from ...apiclient import FakeFlockerClient, Dataset
 from ...control._config import dataset_id_from_name
 
-from ...node.testtools import require_docker_version
-
 from ...restapi.testtools import buildIntegrationTests, APIAssertionsMixin
 
 
+class SimpleCountingProxy(object):
+    """
+    Transparent proxy that counts the number of calls to methods of the
+    wrapped object.
+
+    :ivar _wrapped: Wrapped object.
+    :ivar call_count: Mapping of method name to number of calls.
+    """
+    def __init__(self, wrapped):
+        self._wrapped = wrapped
+        self.call_count = pmap()
+
+    def num_calls(self, name):
+        """
+        Return the number of times the given method was called with given
+        arguments.
+
+        :param name: Method name.
+
+        :return: Number of calls.
+        """
+        return self.call_count.get(name, 0)
+
+    def __getattr__(self, name):
+        method = getattr(self._wrapped, name)
+
+        def counting_proxy(*args, **kwargs):
+            current_count = self.call_count.get(name, 0)
+            self.call_count = self.call_count.set(name, current_count + 1)
+            return method(*args, **kwargs)
+        return counting_proxy
+
+
+@attributes(["clock", "reactor"])
+class AutomaticClock(proxyForInterface(IReactorTime, "clock")):
+    """
+    This is an automatic clock, which schedules itself to advance whenever
+    there is a pending call.
+
+    :ivar Clock clock: The underlying clock that gets all calls proxied to it.
+        This clock is advanced with every call to callLater.
+    :ivar IReactorTime reactor: This ``IReactorTime`` provider is used to
+        provide an alternate context to advance the clock from. It is not
+        realistic to advance the clock from the context that calls
+        ``callLater``, so we require a different context to advance this clock
+        from.
```

Comment:
I feel I must be running low on brain. I'm having trouble understanding this comment and also making suggestions for how it could be better. I guess it's because I don't know what "context to advance the clock from" (and variants) mean.